apiserver/encrypting-data-at-datastore-layer.md

```diff
@@ -0,0 +1,285 @@
+1. A controller that runs storage migration for new write keys (this is what actually causes data to be encrypted under a given write key).  This migration only runs when all API servers are on the same revision to prevent writing of data to etcd with a key that is not observed by all servers.  After the migration is complete, the migration timestamp annotation is set to the current time.
+
+##### Building the EncryptionConfiguration object
+
+The process for building the `EncryptionConfiguration` object is straightforward, though it does involve a number of steps.
+
+1. List all secrets that match the operator's `encryption.apiserver.operator.openshift.io/component` label
+1. Sort them to have ascending key numbers
+1. Iterate over the sorted list and build the current state for each `GroupResource`
+
+State per `GroupResource`:
+
+1. The mapping of secret to actual encryption key
+1. The mapping of actual encryption key to secret
+1. The read, write and migrated state of each secret
+1. The last secret to be migrated
+
+Translation of each `GroupResource`'s state to individual `ResourceConfiguration`s within the complete `EncryptionConfiguration`:
+
+1. Determine the desired write key by iterating over the ordered secrets and short-circuiting on the first matching criteria
+    - If there is a secret with the write-key annotation but not the migrated annotation, it is the desired write key
+    - If there is a secret with the read-key annotation but not the write-key annotation, it is the desired write key
+    - If present, use the last secret to be migrated as the desired write key
+1. Treat all other keys as read keys
+    - Iterate in reverse to order the read keys with highest key number first
+    - If the last secret to be migrated is encountered, short-circuit as all remaining read keys are no longer needed
+1. Use the write key as the first provider and the read keys as later providers in the list of `ProviderConfiguration`s
+1. If there is no write key, use the `IdentityConfiguration` as the first provider, otherwise use it as the last provider
+
+### Risks and Mitigations
+
+1. Deletion of an in-use encryption key will permanently break a cluster
+    - Keys are stored in `openshift-config-managed` which is an immortal namespace
+    - Keys require a two-phase delete via a finalizer as mentioned above
+1. Encrypting a large number of resources may cause a high number of rollouts
+    - Controller #1 above does not update the single complete secret during rollouts of the API server.  This causes it to collapse the addition of many new keys into a smaller set of changes.
```

Comment:
add:

If key rotations spread out over a larger amount of time such that the upper collapsing does not happen, and if this becomes a problem of "continuous redeployments", a key rotation "lock step" mechanism can be added such that the key rotation controller actively collects outstanding key rotations into one point in time to rotate all of them at once. This effectively works against spread out key rotation events.